vxhunter/vxutility/symbol_table.py

```diff
@@ -197,3 +184,125 @@ def get_symtab_bounds(blk, vx_ver, symbol_fn):
+        if sym is None or not sym_valid:
+            symtab = []
+            i += word_size
+            continue
+
+        name_addr = sym['name_addr']
+
+        # the name pointer must be within the bounds of the memory block
+        # if name_addr < blk.start.offset or name_addr >= blk.end.offset:
+        if verify_addresses:
+            if not is_offset_in_current_program(name_addr):
+                symtab = []
+                i += word_size
+                continue
+
+        sym_type = sym['flag']
+
+        # if the symbol type resides in the .text or .data section, make sure it's dest pointer is within the block
+        if sym_type in TEXT_SYM_TYPES[vx_ver] or sym_type in DATA_SYM_TYPES[vx_ver]:
+            dest_addr = sym['dest_addr']
+
+            # if dest_addr < blk.start.offset or dest_addr >= blk.end.offset:
+            if verify_addresses:
+                if not is_offset_in_current_program(dest_addr):
+                    symtab = []
+                    i += word_size
+                    continue
+
+        sys.stdout.write('.')
+        symtab.append(sym)
+
+        # increase the cursor by the symbol size since we have a valid symbol
+        i += sym_st_size
+
+        # if we have 100 consecutive symbols, assume that this is indeed the symbol table
+        if len(symtab) >= SYMTAB_MIN_COUNT:
+            break
+
+    return i, symtab
+
+
+def try_rebase(blk, sym_table, vx_ver):
+    '''Detected that base is probably not right. This will cause symbols to not be generated correctly
+    So try to find the strings table with function symbol names
+    Find the highest pointer to a function name,
+    Subtract the two, AND by 0xFFFF0000 to get an offset that we can apply
+    '''
+    sym_address = None
+
+    # 'VxWorks' occurs just before the table of function name strings. this doesn't need to be perfect
+    # just need something close to the start of the table will do.
+    # Probably need to make this more robust, need to test with a larger set of images.
+    if vx_ver > 5:
+        search_str = '\x00VxWorks\x00'
+    else:
+        search_str = '\x00_GLOBAL_\x00'
+
+    search_result = find_bytes_address(blk.start.offset, search_str)
+    if search_result:
+        sym_address = int(search_result.toString(), 16)
+        # since we're searching with a leading null, we need to increment address by 1.
+        sym_address = sym_address + 1
+
+        max_ptr = 0
+        for sym in sym_table:
+            if sym['name_addr'] > max_ptr:
+                max_ptr = sym['name_addr']
+
+        print_out(
+            'Base doesn\'t look correct... trying to verify base is correct (name_addr)%08x - (Key_string)%08x' % (
+                max_ptr, sym_address))
+        initial_offset = (max_ptr - sym_address) & 0xFFF00000
+        print_out('initial_offset: 0x%08x' % initial_offset)
+
+        offset = fuzzy_search_for_function_names(initial_offset, sym_table)
```

Comment:
I think this method is pretty slick. How well-tested is `0xfff00000` as a mask? I feel like I've seen images aligned to `0x1000` but am not positive.

Reply:
well spotted. And 0x1000 are bases in some images, so the mask is a first pass. I had 0xffff0000 initially but it didn't work for all cases. and even then like you said the 0x1000 weren't caught. So i added the `fuzzy_search_for_function_names` to fine tune the address.

Reply:
Right. All good then.